### ISSUE TITLE
AbstractCarrierInterface::collectRates() has incorrect return signature

### ISSUE THREAD

**Issue by kassner** [magento/magento2#28514]
### Summary

[\Magento\Shipping\Model\Carrier\AbstractCarrierInterface::collectRates](https://github.com/magento/magento2/blob/dc6d0b01fff3988f706f718e16d877bb4b1dea91/app/code/Magento/Shipping/Model/Carrier/AbstractCarrierInterface.php#L24-L31) currently has in its signature that it will return `\Magento\Framework\DataObject`, but technically `\Magento\Shipping\Model\Rate\Result` is used everywhere, and it does not inherit the DataObject, so the signature is, de facto, incorrect.

### Examples

[\Magento\Shipping\Model\Shipping](https://github.com/magento/magento2/blob/dc6d0b01fff3988f706f718e16d877bb4b1dea91/app/code/Magento/Shipping/Model/Shipping.php#L197-L264):

```
        if (!$result) {
            //One shipment for all items.
            $result = $carrier->collectRates($request);
        }

        if (!$result) {
            return $this;
        } elseif ($result instanceof Result) { // use Magento\Shipping\Model\Rate\Result;
            $this->getResult()->appendResult($result, $carrier->getConfigData('showmethod') != 0);
        } else {
            $this->getResult()->append($result);
        }
```

Both append methods expect the result class:

```
    /**
     * Append result received from a carrier.
     *
     * @param Result $result
     * @param bool $appendFailed Append result's errors as well.
     * @return void
     */
    public function appendResult(Result $result, bool $appendFailed): void
    {
        $this->results[] = ['result' => $result, 'appendFailed' => $appendFailed];
    }
```

```
    /**
     * Add a rate to the result
     *
     * @param \Magento\Quote\Model\Quote\Address\RateResult\AbstractResult|\Magento\Shipping\Model\Rate\Result $result
     * @return $this
     */
    public function append($result)
    {
        if ($result instanceof \Magento\Quote\Model\Quote\Address\RateResult\Error) {
            $this->setError(true);
        }
        if ($result instanceof \Magento\Quote\Model\Quote\Address\RateResult\AbstractResult) {
            $this->_rates[] = $result;
        } elseif ($result instanceof \Magento\Shipping\Model\Rate\Result) {
            $rates = $result->getAllRates();
            foreach ($rates as $rate) {
                $this->append($rate);
            }
        }
        return $this;
    }
```

Some Magento shipping modules also don't expect the DataObject to be used:

[Flatrate](https://github.com/magento/magento2/blob/dc6d0b01fff3988f706f718e16d877bb4b1dea91/app/code/Magento/OfflineShipping/Model/Carrier/Flatrate.php#L71-L99):

```
    /**
     * Collect and get rates
     *
     * @param RateRequest $request
     * @return Result|bool
     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
     * @SuppressWarnings(PHPMD.NPathComplexity)
     */
    public function collectRates(RateRequest $request)
    {
        /** @var Result $result */
        $result = $this->_rateResultFactory->create();
...
        return $result;
    }
```

[Freeshipping](https://github.com/magento/magento2/blob/dc6d0b01fff3988f706f718e16d877bb4b1dea91/app/code/Magento/OfflineShipping/Model/Carrier/Freeshipping.php#L84-L128):

```
    /**
     * FreeShipping Rates Collector
     *
     * @param RateRequest $request
     * @return \Magento\Shipping\Model\Rate\Result|bool
     */
    public function collectRates(RateRequest $request)
    {
...
        /** @var \Magento\Shipping\Model\Rate\Result $result */
        $result = $this->_rateResultFactory->create();
...
        return $result;
    }
```

[UPS](https://github.com/magento/magento2/blob/dc6d0b01fff3988f706f718e16d877bb4b1dea91/app/code/Magento/Ups/Model/Carrier.php#L248-L276):

```
    /**
     * Collect and get rates/errors
     *
     * @param RateRequest $request
     * @return Result|Error|bool
     */
    public function collectRates(RateRequest $request)
    {
...
    }
```

[Fedex](https://github.com/magento/magento2/blob/dc6d0b01fff3988f706f718e16d877bb4b1dea91/app/code/Magento/Fedex/Model/Carrier.php#L288-L305):

```
    /**
     * Collect and get rates
     *
     * @param RateRequest $request
     * @return Result|bool|null
     */
    public function collectRates(RateRequest $request)
    {
        if (!$this->canCollectRates()) {
            return $this->getErrorMessage();
        }

        $this->setRequest($request);
        $this->_getQuotes();
        $this->_updateFreeMethodQuote($request);

        return $this->getResult();
    }
```

Basically no class expects `\Magento\Framework\DataObject`, as the AbstractCarrierInterface::collectRates() suggests.

### Proposed solution

Change the method's signature to reflect the real world, but the method is marked with `@api`, so I don't know how to proceed.

**Comment by kassner** [comment]
up

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @gabrieldagama 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-40335` were created

**Issue Available**: @gabrieldagama, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by kassner** [comment]
up

**Comment by engcom-Bravo** [comment]
Hi @kassner,

Since the related PR has been closed and there were no updates since long time.Hence we are closing this issue.

Thanks.
